automation/taskcluster/decision_task_nightly.py

```diff
@@ -3,95 +3,161 @@
+GITHUB_HTTP_REPOSITORY = os.environ.get('MOBILE_GITHUB_HTTP_REPOSITORY')
 HEAD_REV = os.environ.get('MOBILE_HEAD_REV')
 
-def generate_build_task():
-    created = datetime.datetime.now()
-    expires = taskcluster.fromNow('1 year')
-    deadline = taskcluster.fromNow('1 day')
-
-    command = "automation/taskcluster/actions/nightly.sh"
-
-    return {
-        "workerType": 'gecko-focus',
-        "taskGroupId": TASK_ID,
-        "expires": taskcluster.stringDate(expires),
-        "retries": 5,
-        "created": taskcluster.stringDate(created),
-        "tags": {},
-        "priority": "lowest",
-        "schedulerId": "focus-nightly-sched",
-        "deadline": taskcluster.stringDate(deadline),
-        "dependencies": [ TASK_ID ],
-        "routes": [
-            "index.project.mobile.reference-browser.nightly.latest"
-        ],
-        "scopes": [
-            "queue:route:index.project.mobile.reference-browser.nightly.*",
-            "secrets:get:project/mobile/reference-browser/preview-key-store",
+BUILDER = lib.tasks.TaskBuilder(
+    task_id=TASK_ID,
+    owner="skaspari@mozilla.com",
+    source='{}/raw/{}/.taskcluster.yml'.format(GITHUB_HTTP_REPOSITORY, HEAD_REV),
+    scheduler_id=SCHEDULER_ID
+)
+
+
+def generate_build_task(apks):
+    artifacts = {}
+    for apk in apks:
+        artifact = {
+            "type": 'file',
+            "path": "/build/reference-browser/{}".format(apk),
+            "expires": taskcluster.stringDate(taskcluster.fromNow('1 year'))
+        }
+        artifacts["public/{}".format(os.path.basename(apk))] = artifact
+
+    checkout = 'git clone {} && cd reference-browser'.format(GITHUB_HTTP_REPOSITORY)
+
+    return taskcluster.slugId(), BUILDER.build_task(
+        name="(Reference Browser) Build task",
+        description="Build Reference Browser from source code.",
+        command=('cd .. && ' + checkout +
+                 ' && python automation/taskcluster/helper/get-secret.py'
+                 ' -s project/mobile/reference-browser/sentry -k dsn -f .sentry_token'
+                 ' && ./gradlew --no-daemon -PcrashReportingEnabled clean assembleRelease test'),
+        features={
+            "chainOfTrust": True,
+            "taskClusterProxy": True
+        },
+        artifacts=artifacts,
+        scopes=[
             "secrets:get:project/mobile/reference-browser/sentry"
+        ]
+    )
+
+
+def generate_signing_task(build_task_id, apks, is_staging):
+    artifacts = []
+    for apk in apks:
+        artifacts.append("public/{}".format(os.path.basename(apk)))
+
+    routes = []
+
+    scopes = [
+        "project:mobile:reference-browser:releng:signing:format:autograph_apk_reference_browser",
+        "project:mobile:reference-browser:releng:signing:cert:{}".format(
+            'dep-signing' if is_staging else 'release-signing')
+    ]
+
+    index = "index.project.mobile.reference-browser.{}.latest".format(
+        'staging-nightly' if is_staging else 'nightly')
```

Comment:
Not sure if we should leave these staging related checks here. Once this code merges, it'll represent production always. It makes sense to have since we may want to iterate on these again in the future should more features are needed or improvements, but somehow it feels wrong to me. I'll wait to see what @JohanLorenzo thinks of it too but I'm leaning towards removing this and whoever wants to play with staging later, will make the proper adjustements. Depends on the story of staging releases in the mobile world that we'll discuss in Orlando as well.

Reply:
If we wanted, we can probably design a nice way to abstract these so it's cleaner - I do agree that the staging-related checks add complexity to the code.

However, I don't think that it would be a good idea to remove them from the code, because then we have a couple issues:
1. When taking production code to work on, we need to make all the adjustments to make it work with staging - this would become a very manual, potentially arduous process of finding out which scopes/workers haven't been properly staging-ified yet.
2. When taking staging code and putting it back to production, it needs to be patched back - we have the same issue of having to be careful that we caught all the differences between staging and prod
3. This kind of additional work increases the barrier to making improvements, which in turn decreases the fluidity and maintainability of the code. Small quality-of-life changes get deferred, since we don't want to take the time to test them in staging, and we don't want to risk making the changes in production.

I'm pretty confident that putting this behaviour behind a convenient `--staging` switch has a lot of benefits :smile:

Reply:
I agree on keeping `--staging`. Thanks @mitchhentges for explaning the benefits!